pkg/scaling/resolver/gcp_secretmanager_handler.go

```diff
@@ -0,0 +1,139 @@
+	var err error
+
+	podIdentity := vh.gcpSecretsManager.PodIdentity
+	if podIdentity == nil {
+		podIdentity = &kedav1alpha1.AuthPodIdentity{}
+	}
+
+	switch podIdentity.Provider {
+	case "", kedav1alpha1.PodIdentityProviderNone:
+		missingErr := fmt.Errorf("clientSecret is expected when not using a pod identity provider")
+		if vh.gcpSecretsManager.Credentials == nil {
+			return missingErr
+		}
+
+		clientSecretName := vh.gcpSecretsManager.Credentials.ClientSecret.ValueFrom.SecretKeyRef.Name
+		clientSecretKey := vh.gcpSecretsManager.Credentials.ClientSecret.ValueFrom.SecretKeyRef.Key
+		clientSecret := resolveAuthSecret(ctx, client, logger, clientSecretName, triggerNamespace, clientSecretKey, secretsLister)
+
+		if clientSecret == "" {
+			return missingErr
+		}
+
+		gcpCredentials, err := google.CredentialsFromJSON(ctx, []byte(clientSecret), secretmanager.DefaultAuthScopes()...)
+		if err != nil {
+			return fmt.Errorf("failed to get credentials from json, %w", err)
+		}
+
+		vh.gcpSecretsManagerClient, err = secretmanager.NewClient(ctx, option.WithCredentials(gcpCredentials))
+		if err != nil {
+			return fmt.Errorf("failed to create secretmanager client, %w", err)
+		}
+
+		gcpCedentialsMap := make(map[string]interface{})
+		if err := json.Unmarshal([]byte(clientSecret), &gcpCedentialsMap); err != nil {
+			return fmt.Errorf("failed to unmarshal gcp credentials key into a map, %w", err)
+		}
+
+		project, ok := gcpCedentialsMap["project_id"]
+		if !ok {
+			return fmt.Errorf("project_id field is missing in the json credentials")
+		}
+
+		vh.gcpProjectID = project.(string)
+
+	case kedav1alpha1.PodIdentityProviderGCP:
+		if vh.gcpSecretsManagerClient, err = secretmanager.NewClient(ctx); err != nil {
+			return fmt.Errorf("failed to create secretmanager client: %w", err)
+		}
+
+		// Running workload identity outside GKE, we can't use the metadata api
+		// and we need to use the env that it's provided from the hook
+		project, found := os.LookupEnv("CLOUDSDK_CORE_PROJECT")
+		if !found {
+			if project, err = metadata.NewClient(&http.Client{}).ProjectID(); err != nil {
+				return fmt.Errorf("failed to fetch gcp project id: %w", err)
+			}
+		}
+
+		vh.gcpProjectID = project
+
+	default:
+		return fmt.Errorf("gcp secret manager does not support pod identity provider - %v", podIdentity.Provider)
+	}
+
+	return nil
+}
+
+func (vh *GCPSecretManagerHandler) Read(ctx context.Context, secretID, secretVersion string) (string, error) {
+	req := &secretmanagerpb.AccessSecretVersionRequest{
+		Name: fmt.Sprintf("projects/%s/secrets/%s/versions/%s", vh.gcpProjectID, secretID, secretVersion),
+	}
+
+	result, err := vh.gcpSecretsManagerClient.AccessSecretVersion(ctx, req)
+	if err != nil {
+		return "", fmt.Errorf("failed to access the secret %s version %s, %w", secretID, secretVersion, err)
+	}
+
+	crc32c := crc32.MakeTable(crc32.Castagnoli)
+	checksum := int64(crc32.Checksum(result.Payload.Data, crc32c))
+	if checksum != *result.Payload.DataCrc32C {
```

Comment:
nit: the field is marked as [optional](https://github.com/googleapis/google-cloud-go/blob/e8dc540b093ee3d1f4897eca4259d13c13132f24/secretmanager/apiv1/secretmanagerpb/resources.pb.go#L896-L912), might be safer to check for nil
```suggestion
	if result.Payload.DataCrc32C != nil && checksum != *result.Payload.DataCrc32C {
```